File: RecursiveDigitSum.py
'''
We define super digit of an integer x using the following rules:

Given an integer, we need to find the super digit of the integer.

If x  has only 1 digit, then its super digit is x .
Otherwise, the super digit of x is equal to the super digit of the sum of the digits of .
For example, the super digit of 9875 will be calculated as:

	super_digit(9875)   	9+8+7+5 = 29 
	super_digit(29) 	2 + 9 = 11
	super_digit(11)		1 + 1 = 2
	super_digit(2)		= 2  
'''

def superDigit(nums, k):
  # calculate the super digit of n
  # multiply that by k
  # then do our recursion
  # we need to figure out the real number we want a super digit on (super(n) * k)
  # we can call our recursive function again with this number
  superDigit = superDigitRecursive(n)
  combinedSuperDigit = int(superDigit) * k
  
  return superDigitRecursive(str(combinedSuperDigit))

def superDigitRecursive(n):

  # base case, if the length is 1, then we've found our super digit
  # sum everything up
    # traverse over the number and return the sum of what we got
  # call our recursive function with the result

  if len(n) == 1:
    return n
  else:
    return superDigitRecursive(str(sum([int(x) for x in n])))

File: recursiveDigitSum.js
/*
 * Complete the 'superDigit' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. STRING n
 *  2. INTEGER k
 */

function superDigit(n, k) {
  // create our "p"
  // concanate n * k
  // add the digits using recursion
  // call recursiveDigit on p

  let p = n.repeat(k);
  console.log('p', p);
  return recursiveDigit(p);
}
function recursiveDigit(n) {
  // find the base case, if n is a single digit return n
  console.log('recursive n', n);
  if (n.length === 1) return Number(n);
  //traverse over the number and return the sum of what we got. Each digit is str, converto to int
  // total: int
  // call recursive function on the result(as a string)
  let total = 0;
  for (let num = 0; num < n.length; num++) {
    total = total + Number(n[num]);
  }
  return recursiveDigit(total.toString());
}

/**
 * To optimize we will add the n number before concating and then multiply by k:
 * 
 function superDigit(n, k) {
     // dreate our "p"
    // concanate n * k
   // add the digits using recursion
   // call recursiveDigit on p
let super_n = recursiveDigit(n)

let p = super_n * k

return  recursiveDigit(p.toString())

}
 */
